qiskit_machine_learning/connectors/torch_connector.py

```diff
@@ -226,7 +226,12 @@ def __init__(
         self._neural_network = neural_network
 
         self._sparse = sparse
-        self._weights = TorchParam(Tensor(neural_network.num_weights))
+
+        weight = TorchParam(Tensor(neural_network.num_weights))
+        # register param. in graph following PyTorch naming convention
+        self.register_parameter("weight", weight)
```

Comment:
What does this actually register? The variable right above named weight is only in local scope of this method - does this end up registering the property below named weight? Just wondering since self._weights, which was how this was, is what is returned from weight property below - and this was explicitly changed to weight and then assigned to self._weights with this added register_parameter.

Reply:
By registering a parameter under a custom name in a layer (let's call torch connector as a layer in NNs) you expose this parameter as `layer.name` (if I recall correctly), this is quite common in PyTorch models, and add this parameter to a tree of layers/parameters inside PyTorch.
So, when you query model parameters via `named_parameters()` you get them in a list of tuples `(layer_name.param_name, param)`. Usually you see `*.weight` for weights in PyTorch. So here, we'd like to be compatible with PyTorch. Just having an explicit property `weight` is not enough for PyTorch. Either you don't bother registering and in this case the name is your variable name, or you register under a more meaningful name, e.g. to be compliant with built-in layers.

Reply:
I think my concern was more that the `register_parameter` appears to be registering the local variable here `weight = TorchParam(Tensor(neural_network.num_weights))` - maybe that's fine and the register captures it somehow. Or did it really end up registering the property weight from self, since I guess it does use self.weight I guess its the former. I guess with the variable name `weight` the same as the property `weight` whatever was registered was not so clear. Maybe it has to be the local var since otherwise it would have had to have been self.weight. If its the local param or even an instance var I assume it could have been left as it was and done as `self.register_parameter("weight", self._weights)` - did the code need to be changed to make the name the same! Was less than clear to me

Reply:
PyTorch uses an automatic differentiation package called `autograd` which keeps track of all operations performed on torch `Tensor`s using a DAG (Directed Acyclical Graph).  There are different ways of adding nodes to the graph. In this case, we are dealing with `Module` parameters, so we don't only want to keep track of the operations performed on the tensors, we want to register them as parameters for the corresponding layer of the torch `Module`. 

For that, we normally assign an attribute (in this case, `self._weights`) to a`TorchParameter`, a special kind of torch `Tensor`, that gets automatically registered as a `Module` parameter once assigned to an attribute. This is the easy method, but it automatically registers the parameter with the attribute name (`_weights`). Because we don't want that, we propose the alternative of:

1. Declaring a `TorchParameter`, but not assigning it to any `Module` attribute yet (line 230)
2. Registering the parameter into the graph with the desired name (line 232)
3. Assigning the parameter to the attribute after registering it (it will not be re-registered). This operation is tracked by autograd, so any changes in `self._weights` will affect the node registered as `weight` (line 234)

Hopefully this makes sense and adds to Anton's reply.

Reply:
Oh, and in case you were wondering @woodsp-ibm, I thought about doing `self.register_parameter("weight", self._weights)` too, but this option doesn't work because, as explained before, the moment you do `self._weights=TorchParameter(...)`, the parameter gets registered in the graph, and an error is thrown telling you that you are trying to register it twice.

Reply:
Maybe this need to be added as comments then to me I see little logical difference between
```
self._weights = TorchParam(Tensor(neural_network.num_weights))
self.register_parameter("weight", self._weights)
```
and
```
weight = TorchParam(Tensor(neural_network.num_weights))
self.register_parameter("weight", weight)
self._weights = weight
```
except the latter, at least to me, brings in some confusion over name of `weight` when we also have a property called `weight`.  The mere setting of this to some instance varable registers it whereas to a local variable it does not - why does not assigning weight to self._weights in the latter register it again.

Would this would equally as well in regards of things. 
```
weight_tensor = TorchParam(Tensor(neural_network.num_weights))
self.register_parameter("weight", weight_tensor)
```

If its not very intuitive as you say why the code is as it is - for any one that may look at the code in the future to change it for one reason or another I think some explanation around why it is like it is, and that it needs to be that way etc is needed. As you can tell I am struggling with it somewhat as it seems less than logical.